x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java

```diff
@@ -307,20 +289,16 @@ private void ensureOpen() {
                     protected void doRun() {
                         for (SparseFileTracker.Gap gap : gaps) {
                             try {
-                                ensureOpen();
-                                if (readLock.tryLock() == false) {
-                                    throw new AlreadyClosedException("Cache file channel is being evicted, writing attempt cancelled");
+                                if (reference.tryIncRef() == false) {
+                                    throw new AlreadyClosedException("Cache file channel has been released and closed");
                                 }
                                 try {
                                     ensureOpen();
-                                    if (channel == null) {
-                                        throw new AlreadyClosedException("Cache file channel has been released and closed");
-                                    }
-                                    writer.fillCacheRange(channel, gap.start(), gap.end(), gap::onProgress);
-                                    gap.onCompletion();
+                                    writer.fillCacheRange(reference.fileChannel, gap.start(), gap.end(), gap::onProgress);
                                 } finally {
-                                    readLock.unlock();
+                                    reference.decRef();
                                 }
+                                gap.onCompletion();
```

Comment:
Looking at this again I think that `gap.onCompletion();` should be executed within the try/finally block, before the reference is decremented, as completing the gap might trigger the execution of some listeners.

Reply:
I think we have this covered by the fact that we increment the reference by one anyway in `getFileChannelReference` so unless we are in some broken state where the overall listener got failed async (maybe that's impossible even?) the ref counting here is actually redundant.
I actually added the ref counting in `getFileChannelReference` to cover this exact spot so we're not without a reference when completing the listener. (I suppose we could do away with the incrementing in `getFileChannelReference` and decrementing after if we made this change, but it seemed easier to reason about the way it currently is but maybe that's just me?)

Reply:
Yes, sorry, I've been puzzled a minute by the decRef following by the gap's completion instead of the other way around... sorry for the noise you can ignore my comment.